compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -40,13 +39,14 @@ pub fn is_const_evaluatable<'cx, 'tcx>(
         match AbstractConst::new(tcx, def, substs)? {
             // We are looking at a generic abstract constant.
             Some(ct) => {
+                // if the item is trivially substitutable w/o a complex where bound
+                // just substitute it in.
+                if ct.substs.iter().all(|v| matches!(v.unpack(), GenericArgKind::Const{..})) {
+                    return Ok(());
+                }
```

Comment:
this doesn't seem right to me, afaict this allows the following
```rust
fn foo<const N: usize>() {
    let _: [u8; N - 1];
}
```
It looks to me like there is a misunderstanding on either how abstract consts work or a disagreement on how I/we think they should work. 

> In the case where a generic abst. const requires a trivial where bound: `where TypeWithConst<const_fn(N)>:`,
instead of requiring a where bound, just check that only consts are being substituted in to skip over where check.

Why is `const_fn(N)` a trivial where bound here? I don't think that there is a relevant difference between using `N` and `std::mem::size_of::<T>()`, so I don't see why only substituting consts matters here.

Reply:
More likely a misunderstanding on how they work, just because I've not been following them as closely as necessary.
I don't see the need to add a where bound which describes the exact type if it can be automatically derived when there is a where bound with out any rhs, because to me it means the compiler already has all the information and it just needs additional syntax.

I've just not added in types but if you consider a type signature
```
fn foo<T: Ord, const N: usize>(x: T) -> usize {
  if x > SOME_CONST {
    2*N
  } else {
    N
  }
}
```
it would be necessary to propagate the `T: Ord` requirement upwards

Reply:
`[u8; N - 1]:` is not used to say anything relevant about `[u8; N - 1]`, it's only meant as a way to express that `N - 1` will evaluate successfully. We don't yet have any explicit syntax for this but instead just pretty much walk over the function signature and the where clauses and add all found constants.

`ConstEvaluatable` are completely separate from other predicates. `T: Ord` ends up as a `PrediateKind::Trait` so we won't ever look at it in `is_const_evaluatable`.

There also isn't a requirement on a where clause. You can use the where clauses to add additional requirements.
These have to be proven by the caller and are assumed by the function itself to hold.